File: scripts/inventory/Inventory.cs
using System.Collections.Generic;
using System.Linq;

namespace ProiectDeLegume.Scripts.Inventory;

public class Inventory
{
    public double MaxWeight { get; }
    public List<ItemStack> Items { get; } = new();

    public Inventory(double maxWeight = 15.0)
    {
        MaxWeight = maxWeight;
    }

    public double CurrentWeight => Items.Sum(s => s.TotalWeight);
    public double FreeWeight => MaxWeight - CurrentWeight;

    public bool CanAdd(ItemDef def, int count = 1)
    {
        return CurrentWeight + def.Weight * count <= MaxWeight;
    }

    public bool Add(ItemDef def, int count = 1)
    {
        if (!CanAdd(def, count)) return false;

        if (def.Stackable)
        {
            var existing = Items.FirstOrDefault(s => s.Def.Id == def.Id && s.CanAdd());
            if (existing != null)
            {
                int canFit = System.Math.Min(def.MaxStack - existing.Count, count);
                existing.Count += canFit;
                int remainder = count - canFit;
                if (remainder > 0) Items.Add(new ItemStack(def, remainder));
                return true;
            }
        }

        Items.Add(new ItemStack(def, count));
        return true;
    }

    public bool Remove(int index)
    {
        if (index < 0 || index >= Items.Count) return false;
        var stack = Items[index];
        if (stack.Count > 1) stack.Count--;
        else Items.RemoveAt(index);
        return true;
    }
}

File: src/commonMain/kotlin/inventory/Inventory.kt
package inventory

class Inventory(val maxWeight: Double = 15.0) {
    val items = mutableListOf<ItemStack>()

    val currentWeight: Double get() = items.sumOf { it.totalWeight }
    val freeWeight: Double get() = maxWeight - currentWeight

    fun canAdd(itemDef: ItemDef, count: Int = 1): Boolean {
        return currentWeight + itemDef.weight * count <= maxWeight
    }

    /**
     * Add item to inventory. Returns true if successful.
     */
    fun add(itemDef: ItemDef, count: Int = 1): Boolean {
        if (!canAdd(itemDef, count)) return false

        if (itemDef.stackable) {
            val existing = items.find { it.def.id == itemDef.id && it.canAdd() }
            if (existing != null) {
                val canFit = (itemDef.maxStack - existing.count).coerceAtMost(count)
                existing.count += canFit
                val remainder = count - canFit
                if (remainder > 0) {
                    items.add(ItemStack(itemDef, remainder))
                }
                return true
            }
        }

        items.add(ItemStack(itemDef, count))
        return true
    }

    /**
     * Remove one item from a stack. Returns true if successful.
     */
    fun remove(index: Int): Boolean {
        if (index < 0 || index >= items.size) return false
        val stack = items[index]
        if (stack.count > 1) {
            stack.count--
        } else {
            items.removeAt(index)
        }
        return true
    }
}
